src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2218,6 +2218,17 @@ public static bool IsIsOrAsContext(this SyntaxTree syntaxTree, int position, Syn
             // is/as are valid after expressions.
             if (token.IsLastTokenOfNode<ExpressionSyntax>())
             {
+                // Not after lambdas, anonymous methods, or
+                // expressions of type 'void' or 'method group'
+                var expression = token.GetAncestors<ExpressionSyntax>().LastOrDefault();
+                var leftHandBinding = semanticModel.GetSymbolInfo(expression);
+                var symbol = leftHandBinding.GetBestOrAllSymbols().FirstOrDefault();
+                var memberType = symbol?.GetMemberType();
+                if (memberType != null && memberType.SpecialType == SpecialType.System_Void)
+                {
+                    return false;
+                }
+
```

Comment:
I'm thinking of using `var typeInfo = semanticModel.GetTypeInfo(expression);` instead and returning false `if (typeInfo.Type == null)`. However, in the `System.Action b = (() => 0)$$` case, `typeInfo.Type` would equal `System.Action`.

Is that expected? The `typeInfo.Type` for `System.Action b = (() => { })$$` is null.